rrocker-lib/protos/api.proto

```diff
@@ -0,0 +1,106 @@
+syntax = "proto3";
+package rrocker.api;
+
+//Not the greatest design in the world since it doesn't encode
+//exactly which errors can be returned from which function
+//but I'm cutting this corner for simplicity's sake
+enum APIError {
+    ERR_INTERNAL_ERROR = 0; //reserved for truly exceptional situations
+    ERR_TASK_NOT_FOUND = 1;
+    ERR_TASK_STOPPED = 2;
+    ERR_TASK_FINISHED = 3;
+    ERR_BIN_NOT_FOUND = 4;
+}
+
+/// A handle for our task, contains an UUIDv4
+message TaskHandle {
+    string uuid = 1;
+}
+
+/// Memory and CPU constraints for a task. Best effort constraints so 
+/// values above what the host can offer will mean *no* constraints
+message ResourceConstraints {
+    int32 max_cpu = 1; //CPU % of all cores on daemon host
+    int32 max_mem_bytes = 2;  //memory in bytes
+}
+
+/// A message encoding the start task request.
+/// `cmd` is required while `args` and `constraints` are optional
+message StartTaskRequest {
+    string cmd = 1;
+    repeated string args = 2;
+    ResourceConstraints constraints = 3;
+}
+
+/// Task start result, either returns a task handle or error
+message StartTaskReply {
+    oneof result {
+        TaskHandle handle = 1;
+        APIError error = 2;
+    }
+}
+
+/// Task stop result, either a dummy value to indicate it was killed
+/// successfully or an error indicating what when wrong
+message StopTaskReply
+{
+    oneof result {
+        bool ok = 1; //dummy
+        APIError error = 2;
+    }
+}
+
+/// Designates which pipe the program output came from
+enum OutputStream {
+    Stdin = 0; //never used but let's be consistent with the linux api
+    Stdout = 1;
+    Stderr = 2;
+}
+
+/// Task output with a line of output plus which pipe it came from
+message TaskOutput {
+    string line = 1;

```

Comment:
By using a string to represent output, what happens if a particular line of output is very long? Or contains bytes that are not valid as a string?

Reply:
I go into this a bit in the approach document under the rrockerd section:
> Dealing with tty, ptty, terminals and what not is entirely out of scope and as such any program output will be streamed as the raw bytes written to the stdout and stderr pipes.
Furthermore the output will be chunked on newlines and any lines longer than max gRPC packet size will be truncated.

So any line longer than >4MB ([default max recv value](https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/grpc_types.h#L495)) will be truncated.
But yes you're right, I didn't consider that gRPC strings must be valid utf8.

This area is a huge can of worm. For example there are no guarantees a program even writes valid utf8 and it could just spit raw binary out in which case even attempting to chunk it on newlines is futile. 
A pathological case would happen with something as simple as `cat some_big_bin_file`.
I chose newline chunking as that was the simplest and in my opinion the the stream command is meant to be used as log viewer.
As such it made sense to do it the simple way with a lossy utf8 conversion and dropping unprintable characters on the daemon side.
One could also choose to encode each byte as a separate message but that's clearly not scalable. 
Another solution is to then chunk on some max length like 1024 bytes of output but that complicates the problem because now you also need to have a timeout for sending partial chunks which introduces latency.

In a production system I'd probably make some kind of hybrid system that attempts to detect whether the program output is text or binary and then switches between chunking on newlines or number of bytes.

So for this challenge I'll cut a big corner by assuming the output is meant to be human readable and discard invalid utf8.
Of course that means you couldn't do something like `rrocker-cli stream $(rrocker-cli start /bin/cat /dev/sda) > backup.bin` but I think that's a worthwhile tradeoff for this challenge.